pingcap/tiflash
# Make table structure lock not being held when doing learner read

Signed-off-by: JaySon-Huang <tshent@qq.com>

### What problem does this PR solve?

Issue Number: close #815 

Problem Summary:
Now we hold the table's structure lock then do learner read inside `IStorage::read()`. If wait index of learner read blocks, at the same time raft threads need to remove region. Raft thread will try to acquire for table's structure lock. Finally it will block all raft threads, deadlocks between raft and learner read happens.

### What is changed and how it works?

What's Changed:
Abstract learner read routine, and do learner read before acquiring table's structure lock.

How it works:
To avoid this deadlocks, we can do learner read without holding table's structure lock.

### Related changes

- Need to cherry-pick to the release branch 4.0, 3.1

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
  - Create a partition table
  - 1 thread for adding one partition per minute
  - 1 thread for scanning number of partition, drop old partition to keep 30 partitions in total 
  - 1 thread for writing data into partition table
  - 1 thread for reading data from partition table (scanning all partitions)
  - Run for hours to check if any errors return to the client

Side effects

- None

### Release note 

- Fix the issue that concurrency learner read and remove region may meet deadlocks.
